R/f-eval.R

```diff
@@ -72,79 +49,148 @@ f_eval_lhs <- function(f, data = NULL) {
-  arg_env <- new.env(parent = env)
-  arg_env$.env <- complain(env, "Object '%s' not found in environment")
-  arg_env$.data <- complain(data, "Variable '%s' not found in data")
+  # Install fguards and fpromises. Make sure the latter propagate `data`.
+  eval_env$`_F` <- function(f) f
+  eval_env$`_P` <- function(f) f_eval(f, data)
 
-  eval(expr, data, arg_env)
-}
+  # Make unquoting and guarding operators available even when not imported
+  eval_env$`UQ` <- rlang::UQ
+  eval_env$`UQS` <- rlang::UQS
+  eval_env$`UQE` <- rlang::UQE
+  eval_env$`UQF` <- rlang::UQF
 
+  eval_env
+}
 
 #' @rdname f_eval
+#' @param x An object for which you want to find associated data.
+#' @param lookup_msg An error message when your data source is
+#'   accessed inappropriately (by position rather than name).
 #' @export
-find_data <- function(x) UseMethod("find_data")
-
-#' @export
-find_data.NULL <- function(x) list()
+data_source <- function(x, lookup_msg = NULL) {
+  UseMethod("data_source")
+}
 #' @export
-find_data.list <- function(x) x
+data_source.default <- function(x, lookup_msg = NULL) {
+  if (!is_dictionary(x)) {
+    abort("Data source must be a dictionary")
+  }
+  data_source_new(as.list(x), lookup_msg)
+}
 #' @export
-find_data.data.frame <- function(x) x
-
+data_source.data_source <- function(x, lookup_msg = NULL) {
+  if (!is_null(lookup_msg)) {
+    attr(x, "lookup_msg") <- lookup_msg
+  }
+  x
+}
 #' @export
-find_data.default <- function(x) {
-  stop("Do not know how to find data associated with `x`", call. = FALSE)
+data_source.NULL <- function(x, lookup_msg = NULL) {
+  NULL
 }
-
-
-complain <- function(x, message = "object '%s' not found") {
-  if (is.null(x)) {
-    return(NULL)
-  }
-
-  if (is.environment(x)) {
-    x <- clone_env(x)
+#' @export
+data_source.environment <- function(x, lookup_msg = NULL) {
+  lookup_msg <- lookup_msg %||% "Object '%s' not found in environment"
+  if (!identical(x, env_global())) {
+    x <- env_clone(x)
   }
-
-  structure(x, message = message, class = c("complain", class(x)))
+  data_source_new(x, lookup_msg)
+}
+#' @export
+data_source.data.frame <- function(x, lookup_msg = NULL) {
+  lookup_msg <- lookup_msg %||% "Variable '%s' not found in data"
+  data_source_new(x, lookup_msg)
 }
 
-clone_env <- function(x) {
-  list2env(as.list(x, all.names = TRUE), parent = parent.env(x))
+data_source_new <- function(x, lookup_msg) {
+  msg <- lookup_msg %||% "Object '%s' not found in pronoun"
+  class <- c("data_source", class(x))
+  structure(list(src = x), lookup_msg = msg, class = class)
 }
 
 #' @export
-`$.complain` <- function(x, name) {
-  if (!has_name(x, name)) {
-    stop(sprintf(attr(x, "message"), name), call. = FALSE)
+`$.data_source` <- function(x, name) {
+  src <- .subset2(x, "src")
+  if (!has_binding(src, name)) {
+    abort(sprintf(attr(x, "lookup_msg"), name))
   }
-  x[[name]]
+  src[[name]]
 }
-
 #' @export
-`[[.complain` <- function(x, i, ...) {
+`[[.data_source` <- function(x, i, ...) {
```

Comment:
If we're going to do this a little more formally, I think we should also define `length` and `names` methods (and whatever else is needed so the object is consistent).  (At some point I want to write up the methods that you need to override if you override one method - that's going to be quite important if we want to empower people to create new vector classes)